services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java

```diff
@@ -217,6 +217,79 @@ private void patchGoalstateForNeighbor(NetworkConfiguration networkConfig, Unica
+                        subnetStateBuilder.setOperationType(Common.OperationType.INFO);
+                        subnetStateBuilder.setConfiguration(subnetConfigBuilder.build());
+                        multicastGoalState.getGoalStateBuilder().addSubnetStates(subnetStateBuilder.build());
+
+                        // Add subnet to router_state
+                        Router.RouterConfiguration.SubnetRoutingTable.Builder subnetRoutingTableBuilder = Router.RouterConfiguration.SubnetRoutingTable.newBuilder();
+                        subnetRoutingTableBuilder.setSubnetId(subnetEntity.getSubnetId());
+
+                        List<Router.RouterConfiguration.SubnetRoutingTable> subnetRoutingTablesList = new ArrayList<>();
+                        subnetRoutingTablesList.add(subnetRoutingTableBuilder.build());
+
+                        Goalstate.GoalState.Builder goalStateBuilder = multicastGoalState.getGoalStateBuilder();
+                        List<Router.RouterState.Builder> routerStatesBuilders = goalStateBuilder.getRouterStatesBuilderList();
+                        if (routerStatesBuilders != null && routerStatesBuilders.size() > 0) {
+                            subnetRoutingTablesList.addAll(goalStateBuilder.
+                                    getRouterStatesBuilder(0).
+                                    getConfiguration().
+                                    getSubnetRoutingTablesList());
+                            goalStateBuilder.removeRouterStates(0);
+                        }
+
+                        String routerId = subnetEntity.getRouterId();
+                        // If subnet has attached to a router (test scenario #4), we just use the routerId in the subnet.
+                        // Otherwise, we need to get router_state in the networkConfig for test scenario #5.
+                        if (routerId == null) {
+                            List<InternalRouterInfo> internalRouterInfos = networkConfig.getInternalRouterInfos();
+                            for (InternalRouterInfo internalRouterInfo : internalRouterInfos) {
+                                routerId = internalRouterInfo.getRouterConfiguration().getId();
+                                if (routerId != null) break;
+                            }
+                        }
+                        Router.RouterConfiguration.Builder routerConfigBuilder = Router.RouterConfiguration.newBuilder();
+                        routerConfigBuilder.setRevisionNumber(FORMAT_REVISION_NUMBER);
+                        routerConfigBuilder.setHostDvrMacAddress(HOST_DVR_MAC);
+                        routerConfigBuilder.setId(routerId);
+                        routerConfigBuilder.addAllSubnetRoutingTables(subnetRoutingTablesList);
+                        Router.RouterState.Builder routerStateBuilder = Router.RouterState.newBuilder();
+                        routerStateBuilder.setConfiguration(routerConfigBuilder.build());
```

Comment:
Same comment to Router builder.

Reply:
abstract the code in this section to `constructRouterState`